src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,181 @@
     public T set(T value, int index) {
-        return null;
+        checkIndexOutOfBoundsInclusiveSize(index);
+        Node<T> current = getNode(index);
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBoundsInclusiveSize(index);
+        Node<T> current = getNode(index);
+        unlink(current);
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (current.value == object
+                    || (current.value != null
+                    && current.value.equals(object))) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        if (head != null) {
+            head.prev = newNode;
+        }
+        head = newNode;
+        if (tail == null) {
+            tail = head;
+        }
+        size++;
+    }
+
+    private void addLast(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail != null) {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        if (head == null) {
+            head = tail;
+        }
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node == null) {
+            return;
+        }
+        Node<T> prev = node.prev;
+        Node<T> next = node.next;
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        size--;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index > size) {
+            throwIndexOutOfBoundsEx(index);
+        }
+    }
+
+    private void checkIndexOutOfBoundsInclusiveSize(int index) {
+        if (index < 0 || index >= size) {
+            throwIndexOutOfBoundsEx(index);
+        }
+    }
+
+    private void throwIndexOutOfBoundsEx(int index) {
```

Comment:
method that contains only throwing exception is a bad practice. please remove and move the code to the `checkIndexOutOfBoundsInclusiveSize()` method